File: mt-pricing-backend/app/schemas/scraper_sources.py
"""Schemas Pydantic del módulo Scraper Source Builder.

Bloque 1: estructura de la receta (validación del JSONB).
Bloque 2 (Task 10): schemas request/response de la API.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


class RecipeTransform(BaseModel):
    """Transform declarativo de un field. Snippets LLM no entran en F1."""

    op: Literal["regex_capture", "strip_currency", "replace", "map_values", "unit_factor"]
    pattern: str | None = None
    find: str | None = None
    replace_with: str | None = None
    mapping: dict[str, str] | None = None
    factor: float | None = None


class RecipeField(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    selector: str = Field(min_length=1)
    extract: str = Field(default="text")
    type: Literal["str", "float", "int", "currency", "bool"] = "str"
    transform: RecipeTransform | None = None

    @field_validator("extract")
    @classmethod
    def _valid_extract(cls, v: str) -> str:
        if v in ("text", "html") or v.startswith("attr:"):
            return v
        raise ValueError("extract debe ser 'text', 'html', o 'attr:<nombre>'")


class RecipeUrlTemplates(BaseModel):
    search: str | None = None
    pdp: str | None = None
    list: str | None = None
    product: str | None = None


class RecipePagination(BaseModel):
    next_selector: str | None = None
    max_pages: int = Field(default=1, ge=1, le=50)


class Recipe(BaseModel):
    url_templates: RecipeUrlTemplates
    list_item_selector: str | None = None
    pagination: RecipePagination | None = None
    fields: list[RecipeField] = Field(min_length=1)
    anti_bot_hints: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Task 10 — API request / response schemas
# ---------------------------------------------------------------------------
from datetime import datetime
from uuid import UUID


class ScraperSourceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    slug: str = Field(min_length=1, max_length=80, pattern=r"^[a-z0-9-]+$")
    base_url: str = Field(min_length=1)
    destination_profile: Literal["competitor_price", "product_data"]
    fetch_mode: Literal["static", "headless", "stealth"] = "static"
    description: str | None = None
    competitor_brand_id: UUID | None = None


class ScraperSourceRead(BaseModel):
    id: UUID
    name: str
    slug: str
    base_url: str
    description: str | None
    destination_profile: str
    fetch_mode: str
    status: str
    competitor_brand_id: UUID | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RecipeSubmit(BaseModel):
    """Receta enviada por el cliente — se valida contra el schema Recipe."""

    recipe: Recipe


class RecipeRead(BaseModel):
    id: UUID
    source_id: UUID
    version: int
    is_live: bool
    validation_status: str
    has_unapproved_snippet: bool
    recipe: dict[str, Any]

    model_config = {"from_attributes": True}


class ValidateRequest(BaseModel):
    recipe_id: UUID
    test_url: str = Field(min_length=1)


class ValidateResponse(BaseModel):
    status: str
    field_results: dict[str, str]
    records: list[dict[str, Any]]


class ActivateRequest(BaseModel):
    """Body de activación — solo necesita la receta a promover a is_live."""

    recipe_id: UUID

File: mt-pricing-frontend/lib/api/endpoints/scraper-sources.ts
"use client";

import env from "@/lib/env";
import { createSupabaseBrowserClient } from "@/lib/supabase/client";

export type ScraperSourceStatus = "draft" | "testing" | "active" | "disabled" | "degraded";
export type FetchMode = "static" | "headless" | "stealth";
export type DestinationProfile = "competitor_price" | "product_data";
export type ValidationStatus = "unvalidated" | "passing" | "failing";

export interface ScraperSourceRead {
  id: string;
  name: string;
  slug: string;
  base_url: string;
  description: string | null;
  destination_profile: DestinationProfile;
  fetch_mode: FetchMode;
  status: ScraperSourceStatus;
  competitor_brand_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface ScraperSourceCreate {
  name: string;
  slug: string;
  base_url: string;
  destination_profile: DestinationProfile;
  fetch_mode: FetchMode;
  description?: string | null;
}

export interface ScraperSourceUpdate {
  name?: string;
  base_url?: string;
  description?: string | null;
  destination_profile?: DestinationProfile;
  fetch_mode?: FetchMode;
  status?: ScraperSourceStatus;
}

export interface RecipeRead {
  id: string;
  source_id: string;
  version: number;
  is_live: boolean;
  validation_status: ValidationStatus;
  has_unapproved_snippet: boolean;
  recipe: Record<string, unknown>;
  created_at: string;
}

export interface RecipeCreate {
  recipe: Record<string, unknown>;
}

export interface ValidateRequest {
  recipe_id: string;
  test_url: string;
}

export interface ValidateResponse {
  status: string;
  field_results: Record<string, string>;
  records: Record<string, unknown>[];
}

export interface ActivateRequest {
  recipe_id: string;
}

export interface RecipeTransformDef {
  op: "regex_capture" | "strip_currency" | "replace" | "map_values" | "unit_factor";
  pattern?: string;
  find?: string;
  replace_with?: string;
  mapping?: Record<string, string>;
  factor?: number;
}

export interface RecipeFieldDef {
  name: string;
  selector: string;
  extract: string;
  type: "str" | "float" | "int" | "currency" | "bool";
  transform: RecipeTransformDef | null;
}

export interface AnalyzeRequest {
  url: string;
  context?: string | null;
  hint?: string | null;
}

export interface AnalyzeResponse {
  detected_mode: "static" | "headless" | "stealth";
  proposed_source: {
    name: string;
    slug: string;
    base_url: string;
  };
  proposed_recipe: {
    url_templates: { search?: string; pdp?: string; list?: string; product?: string };
    list_item_selector: string | null;
    fields: RecipeFieldDef[];
    anti_bot_hints?: Record<string, unknown>;
  };
  field_confidence: Record<string, number>;
  preview_records: Record<string, unknown>[];
  missing_required: string[];
  warnings: string[];
}

export class ScraperSourcesApiError extends Error {
  public readonly status: number;
  public readonly detail: unknown;

  constructor(status: number, detail: unknown, fallback: string) {
    super(typeof detail === "string" ? detail : fallback);
    this.name = "ScraperSourcesApiError";
    this.status = status;
    this.detail = detail;
  }
}

async function authedFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const supabase = createSupabaseBrowserClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  const headers = new Headers(init.headers);
  if (!headers.has("Content-Type") && init.body) {
    headers.set("Content-Type", "application/json");
  }
  if (session?.access_token) {
    headers.set("Authorization", `Bearer ${session.access_token}`);
  }
  const res = await fetch(`${env.NEXT_PUBLIC_BACKEND_URL}${path}`, {
    ...init,
    headers,
    cache: "no-store",
  });
  if (!res.ok) {
    let detail: unknown;
    try {
      detail = await res.json();
    } catch {
      /* noop */
    }
    throw new ScraperSourcesApiError(res.status, detail, res.statusText);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

export const scraperSourcesApi = {
  list: (): Promise<ScraperSourceRead[]> =>
    authedFetch<ScraperSourceRead[]>("/api/v1/scraper-sources"),

  create: (req: ScraperSourceCreate): Promise<ScraperSourceRead> =>
    authedFetch<ScraperSourceRead>("/api/v1/scraper-sources", {
      method: "POST",
      body: JSON.stringify(req),
    }),

  update: (id: string, req: ScraperSourceUpdate): Promise<ScraperSourceRead> =>
    authedFetch<ScraperSourceRead>(`/api/v1/scraper-sources/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify(req),
    }),

  listRecipes: (sourceId: string): Promise<RecipeRead[]> =>
    authedFetch<RecipeRead[]>(
      `/api/v1/scraper-sources/${encodeURIComponent(sourceId)}/recipes`,
    ),

  createRecipe: (sourceId: string, req: RecipeCreate): Promise<RecipeRead> =>
    authedFetch<RecipeRead>(
      `/api/v1/scraper-sources/${encodeURIComponent(sourceId)}/recipes`,
      {
        method: "POST",
        body: JSON.stringify(req),
      },
    ),

  validate: (sourceId: string, req: ValidateRequest): Promise<ValidateResponse> =>
    authedFetch<ValidateResponse>(
      `/api/v1/scraper-sources/${encodeURIComponent(sourceId)}/validate`,
      {
        method: "POST",
        body: JSON.stringify(req),
      },
    ),

  activate: (sourceId: string, req: ActivateRequest): Promise<ScraperSourceRead> =>
    authedFetch<ScraperSourceRead>(
      `/api/v1/scraper-sources/${encodeURIComponent(sourceId)}/activate`,
      {
        method: "POST",
        body: JSON.stringify(req),
      },
    ),

  analyze: (req: AnalyzeRequest): Promise<AnalyzeResponse> =>
    authedFetch<AnalyzeResponse>("/api/v1/scraper-sources/analyze", {
      method: "POST",
      body: JSON.stringify(req),
    }),
};
